kong/tools/utils.lua

```diff
@@ -1434,21 +1432,28 @@ function _M.sort_by_handler_priority(a, b)
 end
 
 do
-  local counter = 0
+  local get_phase      = ngx.get_phase
+  local ngx_sleep      = ngx.sleep
+  local byte           = string.byte
+  local CHAR_I, CHAR_N = byte("in", 1, 2)
+
+  local counter = YIELD_ITERATIONS
   function _M.yield(in_loop, phase)
     if ngx.IS_CLI then
       return
     end
     phase = phase or get_phase()
-    if phase == "init" or phase == "init_worker" then
+
+    -- phase == "init" or phase == "init_worker"
+    if byte(phase, 1) == CHAR_I and byte(phase, 2) == CHAR_N then
       return
     end
     if in_loop then
-      counter = counter + 1
-      if counter % YIELD_ITERATIONS ~= 0 then
+      counter = counter - 1
+      if counter > 0 then
```

Comment:
I think it is a nice change, I also did a test.

```lua
local random = math.random

local COUNTER = 1e5

local function test_sub()
    local n = COUNTER

    while n > 0 do
        n = n - 1
    end

    return random()
end

local function test_mod()
    local n = 0

    while n % COUNTER ~= 0 do
        n = n + 1
    end

    return random()
end

ngx.update_time()
local start = ngx.now()

test_sub()

ngx.update_time()
print(string.format("test_sub(): %f s", ngx.now() - start))

ngx.update_time()
start = ngx.now()

test_mod()

ngx.update_time()
print(string.format("test_mod(): %f s", ngx.now() - start))
```

```shell
$ resty test.lua
test_sub(): 12.196000 s
test_mod(): 34.310000 s
```

Reply:
Great job!